perf/src/packet.rs

```diff
@@ -13,21 +16,599 @@ use {
 pub const NUM_RCVMMSGS: usize = 64;
 
+/// Representation of a packet used in TPU.
 #[cfg_attr(feature = "frozen-abi", derive(AbiExample))]
-#[derive(Debug, Default, Clone, Serialize, Deserialize)]
-pub struct PacketBatch {
+#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
+pub struct BytesPacket {
+    buffer: Bytes,
+    meta: Meta,
+}
+
+impl BytesPacket {
+    pub fn new(buffer: Bytes, meta: Meta) -> Self {
+        Self { buffer, meta }
+    }
+
+    pub fn from_data<T>(dest: Option<&SocketAddr>, data: T) -> bincode::Result<Self>
+    where
+        T: solana_packet::Encode,
+    {
+        let buffer = BytesMut::with_capacity(PACKET_DATA_SIZE);
+        let mut writer = buffer.writer();
+        data.encode(&mut writer)?;
+        let buffer = writer.into_inner();
+        let buffer = buffer.freeze();
+
+        // We don't use the `size` field of `Meta`here.
+        // Instead, we rely on the length of the buffers. There is no need for
+        // tracking the size manually.
+        // However, removing the `size` field from `Meta` would break ABI and the
+        // `Packet` struct. Maintaining multiple `Meta` implementations would add
+        // more maintenance burden. So let's just set it to MAX.
+        let mut meta = Meta {
+            size: usize::MAX,
+            ..Default::default()
+        };
+        if let Some(dest) = dest {
+            meta.set_socket_addr(dest);
+        }
+
+        Ok(Self { buffer, meta })
+    }
+
+    #[inline]
+    pub fn data<I>(&self, index: I) -> Option<&<I as SliceIndex<[u8]>>::Output>
+    where
+        I: SliceIndex<[u8]>,
+    {
+        if self.meta.discard() {
+            None
+        } else {
+            self.buffer.get(index)
+        }
+    }
+
+    #[inline]
+    pub fn meta(&self) -> &Meta {
+        &self.meta
+    }
+
+    #[inline]
+    pub fn meta_mut(&mut self) -> &mut Meta {
+        &mut self.meta
+    }
+
+    #[inline]
+    pub fn size(&self) -> usize {
+        self.buffer.len()
+    }
+
+    pub fn deserialize_slice<T, I>(&self, index: I) -> bincode::Result<T>
+    where
+        T: serde::de::DeserializeOwned,
+        I: SliceIndex<[u8], Output = [u8]>,
+    {
+        let bytes = self.data(index).ok_or(bincode::ErrorKind::SizeLimit)?;
+        bincode::options()
+            .with_limit(PACKET_DATA_SIZE as u64)
+            .with_fixint_encoding()
+            .reject_trailing_bytes()
+            .deserialize(bytes)
+    }
+
+    pub fn copy_from_slice(&mut self, slice: &[u8]) {
+        self.buffer = Bytes::from(slice.to_vec());
+    }
+
+    #[inline]
+    pub fn as_ref(&self) -> PacketRef<'_> {
+        PacketRef::Bytes(self)
+    }
+
+    #[inline]
+    pub fn as_mut(&mut self) -> PacketRefMut<'_> {
+        PacketRefMut::Bytes(self)
+    }
+}
+
+#[cfg_attr(feature = "frozen-abi", derive(AbiExample, AbiEnumVisitor))]
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub enum PacketBatch {
```

Comment:
Curious about the design here and how I can consume these downstream in banking stage.
If I can be receiving a mixture of inline and bytes packets, then I need to continue to have the internal structures to copy packet data into, and cannot move to doing zero-copy all the time.

Reply:
Yes, that's correct. To be more precise - we won't be able to get rid of all copies in the banking stage as long as the UDP streamer uses `Packet`. Topic for next PRs!